platforms/common/px4_log.cpp

```diff
@@ -45,75 +45,127 @@
-	struct log_message_s log_message;
-	log_message.timestamp = hrt_absolute_time();
-	log_message.severity = 6; //info
+	// we need to advertise with a valid message
+	log_message_s log_message{};
+	log_message.severity = 6; // info
 	strcpy((char *)log_message.text, "initialized uORB logging");
-
+	log_message.timestamp = hrt_absolute_time();
 	orb_log_message_pub = orb_advertise_queue(ORB_ID(log_message), &log_message, log_message_s::ORB_QUEUE_LENGTH);
-
-	if (!orb_log_message_pub) {
-		PX4_ERR("failed to advertise log_message");
-	}
 }
 
-
-__EXPORT void px4_log_modulename(int level, const char *moduleName, const char *fmt, ...)
+__EXPORT void px4_log_modulename(int level, const char *module_name, const char *fmt, ...)
 {
+	static constexpr ssize_t max_length = sizeof(log_message_s::text);
+
 	FILE *out = stdout;
+
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
 	bool use_color = true;
+#endif // PX4_LOG_COLORIZED_OUTPUT
 
-#ifdef __PX4_POSIX
-	out = get_stdout(&use_color);
-#endif
+#if defined(__PX4_POSIX)
+	bool isatty_ = false;
+	out = get_stdout(&isatty_);
 
-#ifndef PX4_LOG_COLORIZED_OUTPUT
-	use_color = false;
-#endif
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+	use_color = isatty_;
+#endif // PX4_LOG_COLORIZED_OUTPUT
+#endif // PX4_POSIX
 
 	if (level >= _PX4_LOG_LEVEL_INFO) {
-		if (use_color) { fputs(__px4_log_level_color[level], out); }
+		char buf[max_length + 1]; // same length as log_message_s::text, but add newline
+		ssize_t pos = 0;
+
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+
+		if (use_color) {
+			pos += sprintf(buf + pos, "%s", __px4_log_level_color[level]);
+		}
+
+#endif // PX4_LOG_COLORIZED_OUTPUT
 
-		fprintf(out, __px4__log_level_fmt __px4__log_level_arg(level));
+		pos += snprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), __px4__log_level_fmt, __px4_log_level_str[level]);
 
-		if (use_color) { fputs(PX4_ANSI_COLOR_GRAY, out); }
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
 
-		fprintf(out, __px4__log_modulename_pfmt, moduleName);
+		if (use_color) {
+			pos += snprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), PX4_ANSI_COLOR_GRAY);
+		}
 
-		if (use_color) { fputs(__px4_log_level_color[level], out); }
+#endif // PX4_LOG_COLORIZED_OUTPUT
+
+		pos += snprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), __px4__log_modulename_pfmt, module_name);
+
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+
+		if (use_color) {
+			pos += snprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), "%s", __px4_log_level_color[level]);
+		}
+
+#endif // PX4_LOG_COLORIZED_OUTPUT
 
 		va_list argptr;
 		va_start(argptr, fmt);
-		vfprintf(out, fmt, argptr);
+		pos += vsnprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), fmt, argptr);
 		va_end(argptr);
 
-		if (use_color) { fputs(PX4_ANSI_COLOR_RESET, out); }
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+
+		if (use_color) {
+			// alway reset color
+			const ssize_t sz = math::min(pos, max_length - (ssize_t)strlen(PX4_ANSI_COLOR_RESET) - (ssize_t)1);
```

Comment:
Do you think the compiler will optimize `strlen(PX4_ANSI_COLOR_RESET)` away or can we help with that in some way?

Reply:
Yes.